specs/core/0_beacon-chain.md

```diff
@@ -1776,12 +1768,13 @@ def process_attestation(state: BeaconState, attestation: Attestation) -> None:
     # Check target epoch, source epoch, source root, and source crosslink
     target_epoch = slot_to_epoch(data.slot)
     assert (target_epoch, data.source_epoch, data.source_root, data.previous_crosslink_root) in {
-        (get_current_epoch(state), state.current_justified_epoch, state.current_justified_root, hash_tree_root(state.current_crosslinks[data.shard])),
-        (get_previous_epoch(state), state.previous_justified_epoch, state.previous_justified_root, hash_tree_root(state.previous_crosslinks[data.shard])),
+        (get_current_epoch(state), state.current_justified_epoch, state.current_justified_root, hash_tree_root(state.current_crosslinks[data.crosslink.shard])),
+        (get_previous_epoch(state), state.previous_justified_epoch, state.previous_justified_root, hash_tree_root(state.previous_crosslinks[data.crosslink.shard])),
     }
 
-    # Check crosslink data root
-    assert data.crosslink_data_root == ZERO_HASH  # [to be removed in phase 1]
+    # Check crosslink
+    assert data.crosslink.epoch = min(slot_to_epoch(data.slot), previous_crosslink.epoch + MAX_CROSSLINK_EPOCHS)
```

Comment:
`previous_crosslink` isn't a declared variable in this scope :(

Reply:
Right—it is either `state.current_crosslinks[data.crosslink.shard]` or `state.previous_crosslinks[data.crosslink.shard]` depending on whether `target_epoch == get_current_epoch(state)` or target_epoch == get_previous_epoch(state)`.

I need to figure out a clean way to present the two cases through `process_attestation`. The `assert (target_epoch, data.source_epoch, data.source_root, data.previous_crosslink_root) in` approach has become unwieldy because of long lines.

Reply:
Bite the bullet and make the two clause expression?

```python
    assert data.target_epoch in (get_current_epoch(state), get_previous_epoch(state)
    if data.target_epoch == get_current_epoch(state):
        assert data.source_epoch == state.current_justified_epoch and data.source_root == state.current_justified_root
        expected_parent_crosslink = state.current_crosslinks[data.shard]
    else:
        assert data.source_epoch == state.previous_justified_epoch and data.source_root == state.previous_justified_root
        expected_parent_crosslink = state.previous_crosslinks[data.shard]       
    # Check crosslink
    assert data.previous_crosslink_root == hash_tree_root(expected_parent_crosslink)
    assert data.crosslink.epoch = min(slot_to_epoch(data.slot), expected_parent_crosslink.epoch + MAX_CROSSLINK_EPOCHS)
```